### ISSUE TITLE
signature: generating presigned_url with elixir library(ExAws)

### ISSUE THREAD

**Issue by mvalitov** [minio/minio#2941]
## Actual behaviour

I use the library [ExAws](https://github.com/CargoSense/ex_aws).
When generating the presigned_url for download file I get this URL:
`http://127.0.0.1:9000/user-1/docs/test/test-test.mp3?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=SDPZD2A3627HX46NM6UU%2F20161015%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20161015T175137Z&X-Amz-Expires=3600&X-Amz-SignedHeaders=host&X-Amz-Signature=faedcb1cfc29103f4e80241a16a762238528ad19aa9142e2afbec51e140027c2`
But when trying to access this URL, I get an error 403:

```
wget "http://127.0.0.1:9000/user-1/docs/test/test-test.mp3?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=SDPZD2A3627HX46NM6UU%2F20161015%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20161015T175137Z&X-Amz-Expires=3600&X-Amz-SignedHeaders=host&X-Amz-Signature=faedcb1cfc29103f4e80241a16a762238528ad19aa9142e2afbec51e140027c2"
The name is too long, 280 chars total.
Trying to shorten...
New name is test-test.mp3?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=SDPZD2A3627HX46NM6UU%2F20161015%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20161015T175137Z&X-Amz-Expires=3600&X-Amz-SignedHeaders=host&X-Amz-Signature=faedcb1cfc29103f4e80.
--2016-10-15 21:04:27--  http://127.0.0.1:9000/user-1/docs/test/test-test.mp3?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=SDPZD2A3627HX46NM6UU%2F20161015%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20161015T175137Z&X-Amz-Expires=3600&X-Amz-SignedHeaders=host&X-Amz-Signature=faedcb1cfc29103f4e80241a16a762238528ad19aa9142e2afbec51e140027c2
Подключение к 127.0.0.1:9000... соединение установлено.
HTTP-запрос отправлен. Ожидание ответа... 403 Forbidden
2016-10-15 21:04:27 ОШИБКА 403: Forbidden.
```

In the browser, I get a text error:

```
<Error>
<Code>SignatureDoesNotMatch</Code>
<Message>
The request signature we calculated does not match the signature you provided. Check your key and signing method.
</Message>
<Key/>
<BucketName/>
<Resource>/user-1/docs/test/test-test.mp3</Resource>
<RequestId>3L137</RequestId>
<HostId>3L137</HostId>
</Error>
```

This URL is wrong?
I tried using both the latest version of the release, and manually compiled version.
## Minio version

Version: 2016-10-14T04:00:39Z
Release-Tag: RELEASE.2016-10-14T04-00-39Z
Commit-ID: 8c59a4166865e65192e91986e8bee329851bbf7c
## System information

Linux marsel 4.4.0-43-generic #63-Ubuntu SMP Wed Oct 12 13:48:03 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux

**Comment by koolhead17** [comment]
Hello @mvalitov thanks for reporting this. Will it be okey to share the code so we can try it at our end?

**Comment by mvalitov** [comment]
Hello @koolhead17.
An [example](https://bitbucket.org/MarsonV/phoenix_minio) of a simple application.
An example of steps to start:
Install dependencies with `mix deps.get`
Create and migrate your database with `mix ecto.create && mix ecto.migrate`
Install Node.js dependencies with `npm install`
Create manually a bucket with name "phoenix", and one blank file with name "sample".
Minio credentials are described in the file config/config.exs.
Start Phoenix endpoint with `mix phoenix.server`
Now you can visit localhost:4000 from your browser.
The home page will also link
![demo_phoenix_minio](https://cloud.githubusercontent.com/assets/8917265/19412938/34667ea6-9329-11e6-9927-639bae42d740.png)

**Comment by harshavardhana** [comment]
```
$ mix ecto.create && mix ecto.migrate
** (Mix) The database for DemoPhoenix.Repo couldn't be created: FATAL (invalid_password): password authentication failed for user "harsha"

01:21:02.997 [error] GenServer #PID<0.197.0> terminating
** (Postgrex.Error) FATAL (invalid_password): password authentication failed for user "harsha"
    (db_connection) lib/db_connection/connection.ex:148: DBConnection.Connection.connect/2
    (connection) lib/connection.ex:622: Connection.enter_connect/5
    (stdlib) proc_lib.erl:247: :proc_lib.init_p_do_apply/3
Last message: nil
State: Postgrex.Protocol
```

Went till here - do you have any pointers on setting up postgres for your use case anything special that we need to consider? 

Thanks.

**Comment by mvalitov** [comment]
If you entered your user password is correct, it is strange. I re-create the [repository](https://bitbucket.org/MarsonV/phoenix_minio), without the use of database.
Install dependencies with `mix deps.get`
Install Node.js dependencies with `npm install`
Create manually a bucket with name "phoenix", and one blank file with name "sample".
Minio credentials are described in the file config/config.exs.
Start Phoenix endpoint with `mix phoenix.server`

**Comment by krishnasrinivas** [comment]
@mvalitov can you check if this fixes it: https://github.com/CargoSense/ex_aws/pull/257

**Comment by mvalitov** [comment]
@krishnasrinivas , unfortunately, the result is the same:

```
wget "http://127.0.0.1:9000/user-1/docs/test/minio.png?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=SDPZD2A3627HX46NM6UU%2F20161016%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20161016T190851Z&X-Amz-Expires=3600&X-Amz-SignedHeaders=host&X-Amz-Signature=7d5c7207c9f5bb449f6815c03ddb61657358b443b642562d966433663c3deb12"
The name is too long, 276 chars total.
Trying to shorten...
New name is minio.png?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=SDPZD2A3627HX46NM6UU%2F20161016%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20161016T190851Z&X-Amz-Expires=3600&X-Amz-SignedHeaders=host&X-Amz-Signature=7d5c7207c9f5bb449f6815c0.
--2016-10-16 22:09:06--  http://127.0.0.1:9000/user-1/docs/test/minio.png?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=SDPZD2A3627HX46NM6UU%2F20161016%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20161016T190851Z&X-Amz-Expires=3600&X-Amz-SignedHeaders=host&X-Amz-Signature=7d5c7207c9f5bb449f6815c03ddb61657358b443b642562d966433663c3deb12
Подключение к 127.0.0.1:9000... соединение установлено.
HTTP-запрос отправлен. Ожидание ответа... 403 Forbidden
2016-10-16 22:09:06 ОШИБКА 403: Forbidden.
```

```
<Error>
<Code>SignatureDoesNotMatch</Code>
<Message>
The request signature we calculated does not match the signature you provided. Check your key and signing method.
</Message>
<Key/>
<BucketName/>
<Resource>/user-1/docs/test/minio.png</Resource>
<RequestId>3L137</RequestId>
<HostId>3L137</HostId>
</Error>
```

**Comment by krishnasrinivas** [comment]
This change made it work:

```
diff --git a/web/controllers/page_controller.ex b/web/controllers/page_controller.ex
index d5e2a51..2fdd1ae 100644
--- a/web/controllers/page_controller.ex
+++ b/web/controllers/page_controller.ex
@@ -4,7 +4,7 @@ defmodule DemoPhoenix.PageController do
   def index(conn, _params) do
     {:ok, url} =
       ExAws.Config.new(:s3)
-      |> ExAws.S3.presigned_url(:http, "phoenix", "sample")
+      |> ExAws.S3.presigned_url(:get, "phoenix", "sample")
     render conn, "index.html", url: url
   end
 end

```

**Comment by mvalitov** [comment]
Thank! It work!

**Comment by mvalitov** [comment]
hello! after updating to the latest development version(install from source), I can not download using the generated presigned_url link.

```
<Error>
<Code>AccessDenied</Code>
<Message>
AWS authentication requires a valid Date or x-amz-date header
</Message>
<Key/>
<BucketName/>
<Resource>/phoenix/sample</Resource>
<RequestId>3L137</RequestId>
<HostId>3L137</HostId>
</Error>
```

Request log:

```
Hypertext Transfer Protocol
     [truncated]GET /phoenix/sample?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=SDPZD2A3627HX46NM6UU%2F20161113%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20161113T142619Z&X-Amz-Expires=3600&X-Amz-SignedHeaders=host&X-Amz-Signature=93d1e
        [ [truncated]Expert Info (Chat/Sequence): GET /phoenix/sample?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=SDPZD2A3627HX46NM6UU%2F20161113%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20161113T142619Z&X-Amz-Expires=3600&X-Amz-SignedHeade]
            [GET /phoenix/sample?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=SDPZD2A3627HX46NM6UU%2F20161113%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20161113T142619Z&X-Amz-Expires=3600&X-Amz-SignedHeaders=host&X-Amz-Signature=93d1ed1dfeb22f3c2]
            [Severity level: Chat]
            [Group: Sequence]
        Request Method: GET
        Request URI [truncated]: /phoenix/sample?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=SDPZD2A3627HX46NM6UU%2F20161113%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20161113T142619Z&X-Amz-Expires=3600&X-Amz-SignedHeaders=host&X-Amz-Signat
        Request Version: HTTP/1.1
    Host: 127.0.0.1:9000\r\n
    Connection: keep-alive\r\n
    Upgrade-Insecure-Requests: 1\r\n
    User-Agent: Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/53.0.2785.143 Safari/537.36\r\n
    Accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8\r\n
    DNT: 1\r\n
    Referer: http://localhost:4000/\r\n
    Accept-Encoding: gzip, deflate, sdch\r\n
    Accept-Language: ru-RU,ru;q=0.8,en-US;q=0.6,en;q=0.4\r\n
    \r\n
    [Full request URI [truncated]: http://127.0.0.1:9000/phoenix/sample?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=SDPZD2A3627HX46NM6UU%2F20161113%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20161113T142619Z&X-Amz-Expires=3600&X-Amz-Signe]
    [HTTP request 1/1]
    [Response in frame: 411]
```

An [example](https://bitbucket.org/MarsonV/phoenix_minio) of a simple application.
An example of steps to start:
Install dependencies with mix deps.get
Install Node.js dependencies with npm install
Create manually a bucket with name "phoenix", and one blank file with name "sample".
Minio credentials are described in the file config/config.exs.
Start Phoenix endpoint with mix phoenix.server

Thank you for any help

**Comment by harshavardhana** [comment]
> <Error>
> <Code>AccessDenied</Code>
> <Message>
> AWS authentication requires a valid Date or x-amz-date header
> </Message>
> <Key/>
> <BucketName/>
> <Resource>/phoenix/sample</Resource>
> <RequestId>3L137</RequestId>
> <HostId>3L137</HostId>
> </Error>

This is because of a recent change which made it stricter, fixing it. @mvalitov  - Thanks for reporting.

**Comment by mvalitov** [comment]
@harshavardhana Thanks! fixed in the master?

**Comment by harshavardhana** [comment]
> @harshavardhana Thanks! fixed in the master?

Not yet.. in a bit stay tuned @mvalitov

**Comment by mvalitov** [comment]
ok, thanks

**Comment by ki0k1** [comment]
@harshavardhana hello! I have same issue in my Elixir project with ExAws.
"The request signature we calculated does not match the signature you provided."

Try build @mvalitov sample project and get "signature not match" too.

minio version: RELEASE.2017-02-16T01-47-30Z

**Comment by kioqq** [comment]
@harshavardhana hello! I have same issue in my Elixir project with ExAws.
"The request signature we calculated does not match the signature you provided."

Try build @mvalitov sample project and get "signature not match" too.

minio version: RELEASE.2017-02-16T01-47-30Z

Bug only with presigned_url , other API works fine.

**Comment by harshavardhana** [comment]
Would you mind providing a test script for us to reproduce and validate ?.

We will fix it . Thanks for reporting.

**Comment by kioqq** [comment]
@harshavardhana Thx for fast answer :) Try build @mvalitov [example](https://github.com/minio/minio/issues/2941#issuecomment-260189703). 


Phoenix v1.2.1
Elixir 1.4.2

`1. git clone https://bitbucket.org/MarsonV/phoenix_minio.git`
`2. cd phoenix_minio`
`3. mix deps.get`
`4. npm install `
`5. set access_key_id/secret_access_key in config/config.exs`
`6. create a bucket with name "phoenix", and one blank file with name "sample".`
`7. mix phoenix.server`
`8. http://localhost:4000`

**Comment by kioqq** [comment]
I try get list_objects and list_buckets from ExAws, works fine!
When i use presigned_url for "phoenix" bucket and "test.png" file.

I get this url:
`http://127.0.0.1:9000/phoenix/test.png?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=1CTN333SZ0VPK44GFEHO%2F20170226%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20170226T005757Z&X-Amz-Expires=3600&X-Amz-SignedHeaders=host&X-Amz-Signature=b10a1f65ed87fd7df0722cb4aa55b8b2da10dea2a38b053097866dbdbd3cf87c`

But if i open url i get this.
<img width="1000" alt="screen shot 2017-02-26 at 3 59 26 am" src="https://cloud.githubusercontent.com/assets/5331145/23336151/2647358c-fbd8-11e6-9af2-324375e79223.png">

**Comment by harshavardhana** [comment]
The issue is at the ex_aws end just saw wrong canonical request generated, but we fixed it upstream - https://github.com/CargoSense/ex_aws/commit/63658792b6c8da247797f7620e59429763c86cba @krishnasrinivas sent a fix long time ago. 

With that fix they made 1.0.0 release, you might have to change the `mix.lock` entry for `ex_aws` to point to this `63658792b6c8da247797f7620e59429763c86cba` commit. 

This fixes the issue, let us know if this fixes the issue for you @ki0k1

**Comment by kioqq** [comment]
@harshavardhana Wow! Big thanks!  It's works for me! 

I check PR's in ex_aws and found it https://github.com/CargoSense/ex_aws/pull/354.
Works great too 👍 

Who have same issue can add this in mix.exs:
"{:ex_aws, git: "https://github.com/hippware/ex_aws.git", branch: "type-fixes"}"

Sorry for bugs in elixir stuff :DD
